### ISSUE TITLE
DAP should use `requirePermissionToResume` to pause any isolates on restart until breakpoints can be set 

### ISSUE THREAD

**Issue by elliette** [dart-lang/sdk#54843]
Copied from https://docs.google.com/document/d/1Ltc4E8R_fxRurnP3M7TXIF_GrOiVCfpQThUBoT__GeY/edit?tab=t.0#heading=h.aofl3jdsu75w:

On a hot-restart or a page reload, there can be a race condition where `runMain` (the hoisted entrypoint to the app) is run before DAP sends back the breakpoints in response to the `kIsolateRunnable event`. This means that any breakpoints set in the app’s entrypoint (the main method of the app) will not be re-established in time to be hit during restart. 

We probably need to start the isolate paused, using DDS’ requirePermissionToResume:  https://github.com/dart-lang/sdk/blob/main/pkg/dds/dds_protocol.md#requirepermissiontoresume *

Then DAP can resume the app once it has sent the breakpoints.

* From @bkonyi : “This is part of the DDS protocol, so DDS needs to be present (you can access this API through package:dds_service_extensions). It's also meant to be used in conjunction with the setClientName RPC, although that probably doesn't matter if we're not having multiple VM service connections from the same tool.”

**Comment by bkonyi** [comment]
@DanTup is this already being done?

**Comment by DanTup** [comment]
@elliette I think you did this in 93556aeee3c155951bee390984fffb86f2ef85c8 (which references this bug, but doesn't close it). Is there anything more needed, or can we close this?

**Comment by DanTup** [comment]
(closing this on the assumption it's done, but please re-open if you think that's not the case)

**Comment by karbiv** [comment]
@elliette  @DanTup 

This commit introduced a bug to DAP implementation.
https://github.com/dart-lang/sdk/commit/93556aeee3c155951bee390984fffb86f2ef85c8

https://github.com/dart-lang/sdk/blob/159bbb6f1b220ad503d08b023803680e54deb024/pkg/dds/lib/src/dap/isolate_manager.dart#L739-L741

Server sends this event when thread exits, note "reason":

```
{
  "seq": 56,
  "type": "event",
  "body": {
    "allThreadsStopped": null,
    "reason": "exit",
    "threadId": 2
  },
  "event": "stopped"
}
```
There's no **"exit"** value for "reason" in DAP specification.

How the bug manifests itself:

In Emacs _dap-mode_ expects any "stopped" event to be for an alive thread, and sets current active thread in UI to that threadId(that's just exited). So it's required to set active thread id manually in order to continue stepping.

Fix:
"stopped" event type should not be sent for an exited thread.

https://microsoft.github.io/debug-adapter-protocol/specification#Events_Stopped

**Comment by DanTup** [comment]
@karbiv 

> "stopped" event type should not be sent for an exited thread.

This event is not the exit itself, but a pause-on-exit. We configure the VM to pause on exit so that we can finish requests that are in-flight even when the program gets to the end. The spec for `reason` includes `| string` so it is allowed to send any string here for the reason. We use "exit" just to indicate the reason that we paused is that we hit the pause-on-exit state.

After we have finished our requests, we will resume the thread and then there will be a thread event for the actual exit. As long as you're getting this event before the actual thread exit event, I think it's working correctly.


> In Emacs dap-mode expects any "stopped" event to be for an alive thread

I'm not sure what "alive" means here. If you mean according to past events, that sounds reasonable. But if you mean you can send another request (like `stackTrace`) and it work successfully, then I think that's an incorrect assumption. Threads can exit at any time, so just because you got a `stopped` event does not mean the thread still exists by the time your next request gets to the debug adapter - you need to handle threads exiting at any time (and that might you not _yet_ have seen their exit events if you're building an outbound request).

**Comment by karbiv** [comment]
>I'm not sure what "alive" means here. 

That "stopped" event with reason value "exit" is sent after a call to `_isolate.kill();`.
So the thread with that threadId does not exist anymore.
But DAP client expects to debug that thread because it has stopped, and sets that threadID as active in a session, UI.

>Threads can exit at any time, so just because you got a stopped event does not mean the thread still exists

Right. Server emits this "thread" event after previous "stopped" event:
```
{
  "seq": 57,
  "type": "event",
  "body": {
    "reason": "exited",
    "threadId": 2
  },
  "event": "thread"
}
```

DAP client uses it to update UI only.  It doesn't change current thread id in a debug session.
Then next "stopped" event flies in and it doesn't update thread id for debug session either:

```
{
  "seq": 58,
  "type": "event",
  "body": {
    "allThreadsStopped": null,
    "reason": "step",
    "threadId": 1
  },
  "event": "stopped"
}
```

It doesn't update thread id for the session because _dap-mode_ updates it only when thread id is NIL in session.
**So, this is a bug of DAP client** in this case. (which is caused by non-standard value in "stopped" DAP event)

**I'll redirect this bug to DAP client then** https://github.com/emacs-lsp/lsp-dart

Even though Dart dev server emits non-standard event, the DAP client should NOT be fragile.

**Comment by DanTup** [comment]
@karbiv yes, I think if the client is behaving badly because of this event, I think that is a bug in the client.

I also don't think `exit` as a `reason` is non-standard, because the spec allows for any `string`. It is certainly a custom value that is not explicit in the spec, but I believe it is valid. VS Code also handles this fine (by just showing "stopped on exit" temporarily before it then handles the thread exiting).

**Comment by DanTup** [comment]
Feel free to CC me in the issue you create if it might be helpful to discuss directly there. Thanks!
